application/common/models/Password.php

```diff
@@ -214,5 +222,64 @@ public function save()
         }
     }
 
+    public function validateNotPublicPassword($attribute)
+    {
+        /*
+         * block passwords provided in https://www.youtube.com/watch?v=3LRnjzA8auk&list=PLu5OsENIeX656zXJ96FCL169WNmvnPveo
+         */
+        $publicPasswords = [
+            'one4amzn',
+            'one4ggle',
+            'one4ebay',
+            '$$Ymh7Hp3dfgQr9L#!>s;',
+            '7startpenguins!snap',
+            'deserty.domes.slide2',
+            'about-slithers-quakely.',
+        ];
 
+        foreach ($publicPasswords as $publicPassword) {
+            if ($this->$attribute == $publicPassword) {
+                $msg = "Ha ha, good one. No, you shouldn't use a password you saw " .
+                    'in a video about creating good passwords.';
+                $this->addError($attribute, \Yii::t('app', $msg));
+            }
+        }
+    }
+
+    /**
+     * Validate a password against the havibeenpwned security breach database.
+     *
+     * Inspired by https://github.com/PayFast/laravel-validate-pwned-password
+     *
+     * @param $attribute
+     */
+    public function validateNotPwned($attribute)
+    {
+        $hash = sha1($this->$attribute);
+        $hashPrefix = substr($hash, 0, 5);
+        $url = 'https://api.pwnedpasswords.com/range/' . $hashPrefix;
+
+        $ch = curl_init();
```

Comment:
Out of curiosity, is there a specific reason we're using `curl` rather than something like Guzzle?

It would also be good to specify a custom User-Agent string (per [the API's docs](https://haveibeenpwned.com/API/v2)), since the acceptable use policy says to the User-Agent should accurately describes the consumer of the API.

Reply:
Regarding `curl` vs Guzzle, I don't have a preference or a good understanding of the pros/cons to make that call. The reference library used `curl` so I didn't mess with it.

I stand guilty of not thoroughly reading the API docs. Thanks for taking time to cover me.